### PR TITLE
Customer friendly preSum

### PR DESCRIPTION
**Pull request by josojo**
Adam's 8) suggestion.

Testplan; unit tests

### CODE REVIEW

File: contracts/EasyAuction.sol
Comment (fedgiac) [inline contracts/EasyAuction.sol]: This is my proposal without using `.previous()`. The reason I don't like `previous` is that it behaves differently from `.next()`, since it exists for deleted orders. This should not be an issue here, since we are on the main chain (this is enforced by the first call to `.next()`). However, I'd still try to avoid `previous` in case future changes break this assumption.

```
        bytes32 previousOrder;
        for (uint256 i = 0; i < iterationSteps; i++) {
            previousOrder = iterOrder;
            iterOrder = sellOrders[auctionId].next(iterOrder);
            (, uint96 buyAmountOfIter, uint96 sellAmountOfIter) =
                iterOrder.decodeOrder();
            sumBidAmount = sumBidAmount.add(sellAmountOfIter);
            if (
                sumBidAmount.mul(buyAmountOfIter) >=
                auctioneerSellAmount.mul(sellAmountOfIter) ||
                iterOrder == IterableOrderedOrderSet.QUEUE_END
            ) {
                sumBidAmount = sumBidAmount.sub(sellAmountOfIter);
                iterOrder = previousOrder;
                break;
            }
        }
```

**Review by fedgiac — APPROVED** [review]
The changes would work as they are. I made a suggestion to remove `.previous()`. since it's a more delicate operation than `.next()` and could lead to issues in case of future changes.

I'd also be interested in the gas costs of these changes. It add quite some cost to each iteration of the loop. I think this might actually be more than a single storage read, so it might make the function able to parse maybe half of the amount of elements compared to before. If auctioneers are worried about having their call to this function reverted, they could just call `settle` to compute the value and use this precalculation with a significantly smaller number of iteration than the maximum possible.

**Comment by josojo** [comment]
> It add quite some cost to each iteration of the loop. I think this might actually be more than a single storage read,

I think there is no additional storage read. It's only memory operations. I think it's not too bad.
Whether the PR is really worth it, I don't know for sure, as this will likely anyways not be used.

File: contracts/EasyAuction.sol
Comment (josojo) [reply contracts/EasyAuction.sol]: Nice, lets take this solution. 
I was under the impression that using a previousOrder might somehow increase the complexity. But for sure it does not. Good!

**Review by fedgiac — APPROVED** [review]
Changes look good. :heavy_check_mark: 

The price of a `SLOAD` is 800 gas. Here for every loop iteration we add two `.mul` (hiding more instructions) and two comparisons. Probably it doesn't reach 800 gas, but I think it still has a visible effect on the number of orders that can be settled.

**Comment by josojo** [comment]
I added an e2e test for direct comparison:

This PR gives
```
Gas usage for verification 205861
```

Old version gives:
```
Gas usage for verification 173478
```
That is 10% less gas. Given that there is no benefit for running the `preSumCalcualtion` and it's expected to be always done by the operator of the exchange, I think we should not merge it.

**Pull request closed without merging**
